File: LC020ValidParentheses.java
package leetcode;

import java.util.Stack;

/**
 * https://leetcode.com/problems/valid-parentheses/
 */
public class LC020ValidParentheses {
    public boolean isValid(String s) {
        if (s == null) return false;
        if (s.length() == 0) return true;
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char character = s.charAt(i);
            switch (character) {
                case '(':
                case '{':
                case '[':
                    stack.push(character);
                    break;
                case ')':
                    if (stack.size() == 0 || stack.peek() != '(') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
                case '}':
                    if (stack.size() == 0 || stack.peek() != '{') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
                case ']':
                    if (stack.size() == 0 || stack.peek() != '[') {
                        return false;
                    } else {
                        stack.pop();
                    }
                    break;
            }
        }
        return stack.size() == 0;
    }

    public static void main(String[] arguments) {
        LC020ValidParentheses solution = new LC020ValidParentheses();
        System.out.println(solution.isValid(")"));
        System.out.println(solution.isValid("()"));
        System.out.println(solution.isValid("()[]{}"));
        System.out.println(solution.isValid("([)]"));
    }
}

File: algorithms/leetcode/lc020_valid_parentheses/lc020_valid_parentheses.py
#!/usr/bin/env python3
# https://leetcode.com/problems/valid-parentheses/

import unittest


class Solution:
    def isValid(self, s: str) -> bool:
        if s is None:
            return False
        if len(s) == 0:
            return True
        stack = []
        for character in s:
            if character in ['(', '{', '[']:
                stack.append(character)
            elif character == ')':
                if len(stack) == 0 or stack[-1] != '(':
                    return False
                stack.pop()
            elif character == '}':
                if len(stack) == 0 or stack[-1] != '{':
                    return False
                stack.pop()
            elif character == ']':
                if len(stack) == 0 or stack[-1] != '[':
                    return False
                stack.pop()
        return len(stack) == 0


class TestCode(unittest.TestCase):
    def test_RE(self):
        self.assertEqual(Solution().isValid(')'), False)

    def test_RB_RE(self):
        self.assertEqual(Solution().isValid('()'), True)

    def test_RB_RE_SB_SE_CB_CE(self):
        self.assertEqual(Solution().isValid('()[]{}'), True)

    def test_RB_SB_RE_SE(self):
        self.assertEqual(Solution().isValid('([)]'), False)

    def test_RB_CB_RE_CE(self):
        self.assertEqual(Solution().isValid('({)}'), False)

    def test_SB_RB_SE_RE(self):
        self.assertEqual(Solution().isValid('[(])'), False)

    def test_RE__SE__CE(self):
        self.assertEqual(Solution().isValid(')'), False)
        self.assertEqual(Solution().isValid(']'), False)
        self.assertEqual(Solution().isValid('}'), False)

    def test_nothing(self):
        self.assertEqual(Solution().isValid(None), False)
        self.assertEqual(Solution().isValid(''), True)
